### PR TITLE
add Container resources to resources list

### PR DESCRIPTION
**Pull request by cigra**
## This PR fixes:
add Container resources to resources list

## Checklist before requesting a review

- [ ] My code follows the style guidelines of this project
- [ ] I have commented on my code, particularly in hard-to-understand areas
- [ ] I have performed a self-review of my code
- [ ] If it is a core feature, I have added thorough tests.
- [ ] New and existing unit tests pass locally with my changes

**Please open the PR against the `dev` branch (Unless the PR contains only documentation changes)**

### CODE REVIEW

File: core/pkg/resultshandling/reporter/v2/reporteventreceiver.go [hunk 1]
```diff
@@ -182,6 +183,55 @@ func (report *ReportEventReceiver) setResults(reportObj *reporthandlingv2.Postur
...
+	if wl == nil {
+		return []reporthandling.Resource{}
+	}
+	containers, err := wl.GetContainers()
+	if err != nil || len(containers) == 0 {
+		return []reporthandling.Resource{}
+	}
+	for idx := range containers {
+		obj := reporthandling.Resource{
+			ResourceID:  uuid.NewString(),
```
Comment (Bezbran) [inline core/pkg/resultshandling/reporter/v2/reporteventreceiver.go:217]: FNV of ParentResourceID + containerName

**Review by Bezbran — CHANGES_REQUESTED** [review]

File: core/pkg/resultshandling/reporter/v2/reporteventreceiver.go [hunk 2]
```diff
@@ -182,6 +183,55 @@ func (report *ReportEventReceiver) setResults(reportObj *reporthandlingv2.Postur
...
+	ApiVersion string //ApiVersion
+	Kind       string //"Container"
+	Metadata   ContainerMetadata
+}
+
+// ContainerMetadata single image object metadata
+type ContainerMetadata struct {
+	Name            string `json:"name"`
+	Namespace       string `json:"namespace"`
+	ParentKind      string `json:"parentKind"`
```
Comment (avrahams) [inline core/pkg/resultshandling/reporter/v2/reporteventreceiver.go]: should we also add parent api version ?

File: core/pkg/resultshandling/reporter/v2/container_test.go [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
...":[{"env":[{"name":"SERVER_PORT","value":"8089"},{"name":"SLEEP_DURATION","value":"1"},{"name":"DEMO_FOLDERS","value":"/app"},{"name":"ARMO_TEST_NAME","value":"auto_attach_deployment"},{"name":"CAA_ENABLE_CRASH_REPORTER","value":"1"}],"image":"quay.io/armosec/demoservice:v25","imagePullPolicy":"IfNotPresent","name":"demoservice","ports":[{"containerPort":8089,"protocol":"TCP"}],"resources":{},"terminationMessagePath":"/dev/termination-log","terminationMessagePolicy":"File"}],"dnsPolicy":"ClusterFirst","restartPolicy":"Always","schedulerName":"default-scheduler","securityContext":{},"terminationGracePeriodSeconds":30}}},"status":{"availableReplicas":1,"conditions":[{"lastTransitionTime":"2021-05-03T13:10:32Z","lastUpdateTime":"2021-05-03T13:10:37Z","message":"ReplicaSet \"demoservice-server-7d478b6998\" has successfully progressed.","reason":"NewReplicaSetAvailable","status":"True","type":"Progressing"},{"lastTransitionTime":"2021-05-03T13:52:58Z","lastUpdateTime":"2021-05-03T13:52:58Z","message":"Deployment has minimum availability.","reason":"MinimumReplicasAvailable","status":"True","type":"Available"}],"observedGeneration":1,"readyReplicas":1,"replicas":1,"updatedReplicas":1}}`,
```
Comment (avrahams) [inline core/pkg/resultshandling/reporter/v2/container_test.go:19]: nicer to use embed and a json file

File: core/pkg/resultshandling/reporter/v2/container.go [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+	ImageHash  string            `json:"imageHash"` //just in kind=pod imageHash:
+	ApiVersion string            `json:"apiVersion"`
+	Kind       string            `json:"kind"`
+	Metadata   ContainerMetadata `json:"metadata"`
+}
+
+type ContainerMetadata struct {
+	Name             string `json:"name"`
+	Namespace        string `json:"namespace"`
+	ParentKind       string `json:"parentKind"`
```
Comment (dwertent) [inline core/pkg/resultshandling/reporter/v2/container.go:25]: Hi, I think we should have an object representing the parent (similar to the `ownerReferences` object)

**Review by dwertent — CHANGES_REQUESTED** [review]

File: core/pkg/resultshandling/reporter/v2/container.go [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+	Parent Metadata `json:"parent"`
+}
+
+type Metadata struct {
+	Kind       string `json:"kind"`
+	Name       string `json:"name"`
+	Namespace  string `json:"namespace,omitempty"`
+	ResourceID string `json:"resourceID,omitempty"`
+	ApiVersion string `json:"apiVersion"`
+}
```
Comment (avrahams) [inline core/pkg/resultshandling/reporter/v2/container.go:34]: Metadata      metav1.ObjectMeta `json:"metadata"`
metav1.TypeMeta `json:",inline"`

File: core/pkg/resultshandling/reporter/v2/reporteventreceiver.go [hunk 6]
```diff
@@ -202,7 +217,9 @@ func (report *ReportEventReceiver) setResources(reportObj *reporthandlingv2.Post
 			return fmt.Errorf("failed to unmarshal resource '%s', reason: %v", resourceID, err)
 		}
 
-		if *counter+len(r) >= MAX_REPORT_SIZE && len(reportObj.Resources) > 0 {
+		containers := rawResourceContainerHandler(*resource)
```
Comment (dwertent) [inline core/pkg/resultshandling/reporter/v2/reporteventreceiver.go:220]: Call this function BEFORE sending resources to the BE (we first collect all the resources we want to send and only then we start sending).
This should happen around [here](https://github.com/kubescape/kubescape/blob/master/core/pkg/resourcehandler/k8sresources.go#L69).

**Review by dwertent — CHANGES_REQUESTED** [review]
@Bezbran I don't understand why this data is generated in kubescape and sent to the BE.
This should be aggregated before inserting the pods to Elastic, simply extract the container out of the pod (exactly what KS is doing...).
Is there a reason this is happening in Kubescape?

**Pull request closed without merging**